File: src/obstacle_detection.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "nav_msgs/msg/occupancy_grid.hpp"
#include "std_msgs/msg/bool.hpp"
#include "visualization_msgs/msg/marker.hpp"
#include <vector>
#include <cmath>
#include <opencv2/opencv.hpp>
#include "tf2/LinearMath/Quaternion.h"
#include "tf2/LinearMath/Matrix3x3.h"


class ObstacleDetection : public rclcpp::Node {
public:
    ObstacleDetection()
    : Node("obstacle_detection") {
        // Inicjalizacja subskrybentów
        auto custom_qos = rclcpp::QoS(rclcpp::KeepLast(5)).best_effort().durability_volatile();
        lidar_subscription_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/sensing/lidar/scan", custom_qos,
            std::bind(&ObstacleDetection::lidar_callback, this, std::placeholders::_1));
        map_subscription_ = this->create_subscription<nav_msgs::msg::OccupancyGrid>(
            "/map", 10,
            std::bind(&ObstacleDetection::map_callback, this, std::placeholders::_1));
        pose_subscription_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
            "/localization/cartographer/pose", 10,
            std::bind(&ObstacleDetection::pose_callback, this, std::placeholders::_1));

        // Inicjalizacja publisherów
        scan_publisher_ = this->create_publisher<sensor_msgs::msg::LaserScan>("/filtered_lidar_scan", 10);
        marker_publisher_ = this->create_publisher<visualization_msgs::msg::Marker>("visualization_marker", 10);
        obstacle_publisher_ = this->create_publisher<std_msgs::msg::Bool>("/obstackle_detected", 10);
        map_publisher_ = this->create_publisher<nav_msgs::msg::OccupancyGrid>("/modified_map", rclcpp::SensorDataQoS());

        // Ustawienia początkowe
        current_map_ = nullptr;
        current_pose_ = nullptr;
    }

private:
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr lidar_subscription_;
    rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr map_subscription_;
    rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr pose_subscription_;
    rclcpp::Publisher<sensor_msgs::msg::LaserScan>::SharedPtr scan_publisher_;
    rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr marker_publisher_;
    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr obstacle_publisher_;
    rclcpp::Publisher<nav_msgs::msg::OccupancyGrid>::SharedPtr map_publisher_;

    nav_msgs::msg::OccupancyGrid::SharedPtr current_map_;
    geometry_msgs::msg::PoseStamped::SharedPtr current_pose_;
    float map_resolution_;
    geometry_msgs::msg::Pose map_origin_;

    void lidar_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg)
    {
        if (!current_map_ || !current_pose_)
            return;

        auto filtered_scan = *msg; // Direct copy of the message

        double pose_yaw = get_yaw_from_quaternion(
            current_pose_->pose.orientation.x,
            current_pose_->pose.orientation.y,
            current_pose_->pose.orientation.z,
            current_pose_->pose.orientation.w);

        double lidar_x = current_pose_->pose.position.x + 0.35 * cos(pose_yaw);
        double lidar_y = current_pose_->pose.position.y + 0.35 * sin(pose_yaw);

        visualization_msgs::msg::Marker marker;
        marker.header.frame_id = "map";
        marker.type = marker.POINTS;
        marker.action = marker.ADD;
        marker.pose.orientation.w = 1.0;
        marker.scale.x = 0.04;
        marker.scale.y = 0.04;
        marker.color.a = 1.0;
        marker.color.r = 1.0;
        marker.color.g = 1.0;
        marker.color.b = 0.0;

        for (size_t i = 0; i < filtered_scan.ranges.size(); ++i) {
            float range = filtered_scan.ranges[i];
            if (std::isinf(range)) continue;

            double angle = msg->angle_min + i * msg->angle_increment;
            double x = lidar_x + range * cos(angle + pose_yaw);
            double y = lidar_y + range * sin(angle + pose_yaw);

            int map_x = static_cast<int>((x - map_origin_.position.x) / map_resolution_);
            int map_y = static_cast<int>((y - map_origin_.position.y) / map_resolution_);

            geometry_msgs::msg::Point p;
            p.x = x;
            p.y = y;
            p.z = 0.0;
            marker.points.push_back(p);

            bool is_occupied = false;
            if (map_x >= 0 && map_x < current_map_->info.width &&
                map_y >= 0 && map_y < current_map_->info.height) {
                int idx = map_y * current_map_->info.width + map_x;
                is_occupied = (current_map_->data[idx] > 10);
            }

            if (is_occupied) {
                filtered_scan.ranges[i] = std::numeric_limits<float>::infinity();
            }
        }

        scan_publisher_->publish(filtered_scan);
        marker_publisher_->publish(marker);

        std_msgs::msg::Bool msg_bool;
        msg_bool.data = std::any_of(filtered_scan.ranges.begin(), filtered_scan.ranges.end(),
                                    [](float range) { return range < 1.7; });

        obstacle_publisher_->publish(msg_bool);
    }

    void map_callback(const nav_msgs::msg::OccupancyGrid::SharedPtr msg) {
        process_occupancy_grid(*msg);
        current_map_ = msg;
    }

    void pose_callback(const geometry_msgs::msg::PoseStamped::SharedPtr msg) {
        
        current_pose_ = msg;
    }

    void process_occupancy_grid(const nav_msgs::msg::OccupancyGrid& occupancy_grid) {
        // Get grid dimensions
        int width = occupancy_grid.info.width;
        int height = occupancy_grid.info.height;

        // Convert data to a cv::Mat
        cv::Mat grid(height, width, CV_8SC1, const_cast<int8_t*>(occupancy_grid.data.data()));

        // Label connected components using 8-connectivity
        cv::Mat labels, stats, centroids;
        int num_features = cv::connectedComponentsWithStats(grid, labels, stats, centroids, 8, CV_32S);

        // Remove clusters with size between 10 and 20
        for (int i = 1; i < num_features; ++i) {
            int cluster_size = stats.at<int>(i, cv::CC_STAT_AREA);
            if (cluster_size > 10 && cluster_size < 20) {
                // Set label areas to 0
                grid.setTo(0, labels == i);
            }
        }

        // Convert cv::Mat back to vector
        std::vector<int8_t> modified_data(grid.begin<int8_t>(), grid.end<int8_t>());

        // Publish modified occupancy grid
        auto occupancy_grid_new = std::make_shared<nav_msgs::msg::OccupancyGrid>(occupancy_grid);
        occupancy_grid_new->data = modified_data;
        map_publisher_->publish(*occupancy_grid_new);
    }

    double get_yaw_from_quaternion(double x, double y, double z, double w) {
        tf2::Quaternion quat(x, y, z, w);
        tf2::Matrix3x3 mat(quat);
        double roll, pitch, yaw;
        mat.getRPY(roll, pitch, yaw);
        return yaw;
    }


};

int main(int argc, char ** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ObstacleDetection>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: scripts/obstacle_detection.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
import math
import numpy as np

from std_msgs.msg import String, Bool 
from sensor_msgs.msg import LaserScan 
from nav_msgs.msg import OccupancyGrid
from geometry_msgs.msg import PoseStamped

from visualization_msgs.msg import Marker
from geometry_msgs.msg import Point

from scipy.ndimage import label


from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy, HistoryPolicy

custom_qos = QoSProfile(
    reliability=ReliabilityPolicy.BEST_EFFORT,
    durability=DurabilityPolicy.VOLATILE,
    history=HistoryPolicy.KEEP_LAST,
    depth=5
)

class ObstacleDetection(Node):

    def __init__(self):
        super().__init__('obstacle_detection')
        self.subscription = self.create_subscription(
            LaserScan,
            '/sensing/lidar/scan',
            self.lidar_callback,
            qos_profile=custom_qos)
        self.publisher_bool = self.create_publisher(Bool, '/obstackle_detected', 10)
        self.subscription  # prevent unused variable warning
        self.subscription_map = self.create_subscription(
            OccupancyGrid,
            '/map',
            self.map_callback,
            10)
        self.subscription_pose = self.create_subscription(
            PoseStamped,
            '/localization/cartographer/pose',
            self.pose_callback,
            10)
        self.publisher_ = self.create_publisher(LaserScan, '/filtered_lidar_scan', 10)
        self.marker_pub = self.create_publisher(Marker, 'visualization_marker', 10)
        self.pub = self.create_publisher(OccupancyGrid, '/modified_map',  0)

        self.current_map = None
        self.current_pose = None


 
    def euler_from_quaternion(self, x, y, z, w):
        """
        Convert a quaternion into euler angles (roll, pitch, yaw)
        roll is rotation around x in radians (counterclockwise)
        pitch is rotation around y in radians (counterclockwise)
        yaw is rotation around z in radians (counterclockwise)
        """
        t0 = +2.0 * (w * x + y * z)
        t1 = +1.0 - 2.0 * (x * x + y * y)
        roll_x = math.atan2(t0, t1)
    
        t2 = +2.0 * (w * y - z * x)
        t2 = +1.0 if t2 > +1.0 else t2
        t2 = -1.0 if t2 < -1.0 else t2
        pitch_y = math.asin(t2)
    
        t3 = +2.0 * (w * z + x * y)
        t4 = +1.0 - 2.0 * (y * y + z * z)
        yaw_z = math.atan2(t3, t4)
    
        return roll_x, pitch_y, yaw_z # in radians

    def get_yaw_from_quaternion(self, x, y, z, w):
        _, _, yaw = self.euler_from_quaternion(x, y, z, w)
        return yaw

    def map_callback(self, msg):
        self.process_occupancy_grid(msg)
        self.current_map = msg
        
        

    def pose_callback(self, msg):
        self.current_pose = msg

    def process_occupancy_grid(self, occupancy_grid):
        # Konwersja danych mapy do tablicy numpy
        width = occupancy_grid.info.width
        height = occupancy_grid.info.height
        grid = np.array(occupancy_grid.data, dtype=np.int8).reshape((height, width))

        # Label connected components using 8-connectivity
        structure = np.ones((3, 3), dtype=np.int8)  # 8-connectivity
        labeled, num_features = label(grid > 0, structure=structure)

        # Remove clusters with size between 10 and 20
        for i in range(1, num_features + 1):
            cluster_size = (labeled == i).sum()
            if 10 < cluster_size < 20:
                grid[labeled == i] = 0
        # Zamiana tablicy numpy z powrotem na listę
        occupancy_grid_new = OccupancyGrid()
        occupancy_grid_new = occupancy_grid
        occupancy_grid_new.data = grid.astype(np.int8).flatten().tolist()

        # Publikowanie zmodyfikowanej mapy
        self.pub.publish(occupancy_grid_new)
        self.current_map = occupancy_grid_new
        


    def lidar_callback(self, msg):
        if self.current_map is None or self.current_pose is None:
            return

        # Kopia danych LiDAR
        filtered_scan = LaserScan()
        filtered_scan.header = msg.header
        filtered_scan.angle_min = msg.angle_min
        filtered_scan.angle_max = msg.angle_max
        filtered_scan.angle_increment = msg.angle_increment
        filtered_scan.time_increment = msg.time_increment
        filtered_scan.scan_time = msg.scan_time
        filtered_scan.range_min = msg.range_min
        filtered_scan.range_max = msg.range_max
        filtered_scan.ranges = msg.ranges[:]
        filtered_scan.intensities = msg.intensities[:]
        
        # Przygotowanie danych do transformacji
        map_resolution = self.current_map.info.resolution
        map_origin = self.current_map.info.origin
        pose_x = self.current_pose.pose.position.x
        pose_y = self.current_pose.pose.position.y
        quat_x = self.current_pose.pose.orientation.x
        quat_y = self.current_pose.pose.orientation.y
        quat_z = self.current_pose.pose.orientation.z
        quat_w = self.current_pose.pose.orientation.w
        pose_yaw = self.get_yaw_from_quaternion(quat_x, quat_y, quat_z, quat_w)

        marker = Marker()
        marker.header.frame_id = "/map"
        marker.type = marker.POINTS
        marker.action = marker.ADD
        marker.pose.orientation.w = 1.0
        marker.scale.x = 0.04  # rozmiar punktów
        marker.scale.y = 0.04
        marker.color.a = 1.0  # Przezroczystość
        marker.color.r = 1.0  # Kolor czerwony
        marker.color.g = 1.0
        marker.color.b = 0.0

        lidar_offset_x = 0.35  # przesunięcie LiDAR od środka robota w metrach
        lidar_offset_y = 0.0  # brak przesunięcia w osi y

        # Oblicz współrzędne przesuniętego LiDAR
        lidar_x = pose_x + lidar_offset_x * np.cos(pose_yaw) 
        lidar_y = pose_y + lidar_offset_x * np.sin(pose_yaw) 



        # Filtracja skanu
        for i in range(len(filtered_scan.ranges)):
            range_val = filtered_scan.ranges[i]
            if range_val == float('Inf'):
                continue
            angle = msg.angle_min + i * msg.angle_increment
            # Przekształcenie do układu mapy
            # print(f'angle {angle} range_val {range_val} pose_yaw {pose_yaw}')
            # jak wjezdamy w sciane to wtedy jest nan
            if range_val != range_val:
                continue
            x = lidar_x + range_val * np.cos(angle + pose_yaw)
            y = lidar_y + range_val * np.sin(angle + pose_yaw)
            
            # Przekształcenie do współrzędnych mapy
            # print(f'x {x} map_origin.position.x) {map_origin.position.x}) / map_resolution/ {map_resolution}')
            map_x = int((x - map_origin.position.x) / map_resolution)
            map_y = int((y - map_origin.position.y) / map_resolution)
            # print(f'map_x {map_x} map_y {map_y}  x/ {(x - map_origin.position.x) / map_resolution}')

            p = Point()
            p.x = map_x * self.current_map.info.resolution + self.current_map.info.origin.position.x
            p.y = map_y * self.current_map.info.resolution + self.current_map.info.origin.position.y
            p.z = 0.0
            marker.points.append(p)
            
            # Sprawdzenie czy punkt znajduje się na ścianie
            # if 0 <= map_x < self.current_map.info.width and 0 <= map_y < self.current_map.info.height:
            #     index = map_y * self.current_map.info.width + map_x
            #     if self.current_map.data[index] > 10:  # założenie, że wartości >50 oznaczają zajętość
            #         filtered_scan.ranges[i] = float('Inf')  # ustawienie zasięgu na nieskończoność

            is_occupied = (
                0 <= map_x < self.current_map.info.width and 0 <= map_y < self.current_map.info.height and
                self.current_map.data[map_y * self.current_map.info.width + map_x] > 10  and 
                self.current_map.data[map_y * self.current_map.info.width + map_x] == -1# Check current cell
            )
            for neighbor_x, neighbor_y in [
                (-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)
            ]:
                neighbor_map_x = map_x + neighbor_x
                neighbor_map_y = map_y + neighbor_y
                if 0 <= neighbor_map_x < self.current_map.info.width and 0 <= neighbor_map_y < self.current_map.info.height:
                    is_occupied |= self.current_map.data[neighbor_map_y * self.current_map.info.width + neighbor_map_x] > 10


            

            if is_occupied:
                filtered_scan.ranges[i] = float('Inf')  # Set range to infinity if occupied

                # to ma byc odszumianie pojedynczych skanow  

        # for i in range(len(filtered_scan.ranges)):
        #     range_val = filtered_scan.ranges[i]
        #     if range_val == float('Inf'):  # Skip already marked points
        #         continue
        #     angle = msg.angle_min + i * msg.angle_increment

        #     # Transform to map coordinates (same as before)
        #     x = pose_x + range_val * np.cos(angle + pose_yaw)
        #     y = pose_y + range_val * np.sin(angle + pose_yaw)
        #     map_x = int((x - map_origin.position.x) / map_resolution)
        #     map_y = int((y - map_origin.position.y) / map_resolution)

        #     # Check if the point has any valid neighbors
        #     has_valid_neighbors = False
        #     for neighbor_x, neighbor_y in [
        #         (-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)
        #     ]:
        #         neighbor_map_x = map_x + neighbor_x
        #         neighbor_map_y = map_y + neighbor_y
        #          # Check if neighbor is within bounds before accessing
        #         if 0 <= neighbor_map_x < map_width and 0 <= neighbor_map_y < map_height:
        #             if isolated_scan.ranges[neighbor_map_y * self.current_map.info.width + neighbor_map_x] != float('Inf'):
        #                 has_valid_neighbors = True
        #                 break
        #         # if 0 <= neighbor_map_x < self.current_map.info.width and 0 <= neighbor_map_y < self.current_map.info.height:
        #         #     if filtered_scan.ranges[neighbor_map_y * self.current_map.info.width + neighbor_map_x] != float('Inf'):
        #         #         has_valid_neighbors = True
        #         #         break

        #     # Mark isolated points as occupied in the isolated scan
        #     if not has_valid_neighbors:
        #         filtered_scan.ranges[i] = float('Inf')

        
        # Publikacja przefiltrowanego skanu
        self.publisher_.publish(filtered_scan)

        self.marker_pub.publish(marker)
        obstacle_bool = False
        for i in range(len(filtered_scan.ranges)):
            distance = filtered_scan.ranges[i]
            if distance < 1.7:  
                self.get_logger().info(f'Obstacle detected at {distance} meters at angle {filtered_scan.angle_min + i * filtered_scan.angle_increment}')
                obstacle_bool = True

        # self.get_logger().info('I heard' )
        msg_bool = Bool()
        msg_bool.data = obstacle_bool
        self.publisher_bool.publish(msg_bool)
        # self.get_logger().info(f'Publishing: {msg.data}')


def main(args=None):
    rclpy.init(args=args)

    obstacle_detection = ObstacleDetection()

    rclpy.spin(obstacle_detection)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    obstacle_detection.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()